Add documentation for QuadraticProgramming. (#109)
google/jaxopt: docs/quadratic_programming.rst

Quadratic programming
=====================

> 

Reviewer: This section is concerned with minimizing quadratic functions subject to equality and/or inequality constraints, also known as quadratic programming.

The different specificities of the solvers are summarized in the table below.  
The best choice will depend on the usage.  
  

> - *jit*: the algorithm can used with jit or vmap, on GPU/TPU.

Reviewer: ```suggestion
- *jit*: the algorithm can be used with jit or vmap, on GPU/TPU.
```

- *precision*: accuracy expected when the solver succeeds to converge.
- *stability*: capacity to handle badly scaled problems and matrices with poor conditioning.  
- *speed*: typical speed on big instances to each its maximum accuracy.
- *input format*: see subsections below.

.. list-table:: Quadratic Solvers
   :widths: 45, 15, 20, 15, 15, 15, 22
   :header-rows: 1

   * - Name
     - jit
     - matvec
     - precision
     - stability
     - speed
     - input format
   * - :class:`EqualityConstrainedQP`
     - yes
     - yes
     - ++
     - \+
     - +++
     - (Q, c), (A, b)
   * - :class:`CvxpyQP`
     - no
     - no
     - +++
     - +++
     - \+
     - (Q, c), (A, b), (G, h)
   * - :class:`OSQP`
     - yes
     - yes
     - \+
     - ++
     - ++
     - (Q, c), (A, b), (G, h)
   * - :class:`BoxOSQP`
     - yes
     - yes
     - \+
     - ++
     - ++
     - (Q, c), A, (l, u)
   * - :func:`linear_solve.solve_cg`
     - yes
     - yes
     - +++
     - +++
     - +++
     - Q, c

This table is given as rule of thumb only; on some particular instances
some solvers may behave unexpectedly better than others.

> In front of difficulties, we suggest to test different combinations of

Reviewer: ```suggestion
In case of difficulties, we suggest to test different combinations of
```

algorithms, ``maxiter`` and ``tol`` values.  
  
.. warning::

  Those algorithms are guaranteed to converge on **convex problems** only.

>   Hence, the matrix :math:`Q` *must* be Positive Semi Definite. 

Reviewer: ```suggestion
  Hence, the matrix :math:`Q` *must* be positive semi-definite (PSD). 
```

Equality-constrained QPs
------------------------

The problem takes the form:

.. math::

    \min_{x} \frac{1}{2} x^\top Q x + c^\top x \textrm{ subject to } A x = b

This class is optimized for QPs with equality constraints only: it supports jit, pytrees and matvec.

.. autosummary::
  :toctree: _autosummary

    jaxopt.EqualityConstrainedQP

Example::

  from jaxopt import EqualityConstrainedQP

  Q = 2 * jnp.array([[2.0, 0.5], [0.5, 1]])
  c = jnp.array([1.0, 1.0])
  A = jnp.array([[1.0, 1.0]])
  b = jnp.array([1.0])

  qp = EqualityConstrainedQP()

>   sol, _ = qp.run(params_obj=(Q, c), params_eq=(A, b)).params

Reviewer: I think `, _` is not needed since you already uses `.params`

  print(sol.primal)
  print(sol.dual_eq)

Ill-posed problems
~~~~~~~~~~~~~~~~~~

This solver is the fastest for well posed problems, but can behave poorly on badly scaled matrices,
or with redundant constraints.

> If the solver struggles to converge, set ``refine_regularization`` and ``refine_maxiter``::

Reviewer: Maybe mention that it uses iterative refinement and link to the wikipedia article.

  from jaxopt._src.eq_qp import EqualityConstrainedQP

  Q = 2 * jnp.array([[3000., 0.5], [0.5, 1]])
  c = jnp.array([1.0, 1.0])
  A = jnp.array([[1.0, 1.0]])
  b = jnp.array([1.0])

  qp = EqualityConstrainedQP(tol=1e-5, refine_regularization=3., refine_maxiter=50)

>   sol, _ = qp.run(params_obj=(Q, c), params_eq=(A, b)).params

Reviewer: Same

  print(sol.primal)
  print(sol.dual_eq)
  print(qp.l2_optimality_error(sol, params_obj=(Q, c), params_eq=(A, b)))

General QPs
-----------

The problem takes the form:

.. math::

    \min_{x} \frac{1}{2} x^\top Q x + c^\top x \textrm{ subject to } A x = b, G x \le h

CvxpyQP
~~~~~~~

The wrapper over Cvxpy is a high precision solver that runs in ``float64``.
However it is not jittable, and does not support matvec and pytrees.

.. autosummary::
  :toctree: _autosummary

    jaxopt.CvxpyQP

Example::

  from jaxopt import CvxpyQP

  Q = 2 * jnp.array([[2.0, 0.5], [0.5, 1]])
  c = jnp.array([1.0, 1.0])
  A = jnp.array([[1.0, 1.0]])
  b = jnp.array([1.0])
  G = jnp.array([[-1.0, 0.0], [0.0, -1.0]])
  h = jnp.array([0.0, 0.0])

  qp = CvxpyWrapper()

>   sol, _ = qp.run(params_obj=(Q, c), params_eq=(A, b), params_ineq=(G, h)).params

Reviewer: Same

  print(sol.primal)
  print(sol.dual_eq)
  print(sol.dual_ineq)

> Only one of ``params_eq`` and ``params_ineq`` are is mandatory.

Reviewer: ```suggestion
It is also possible to specify only equality constraints or only inequality constraints by setting ``params_eq`` or ``params_ineq`` to ``None``.
```

OSQP
~~~~

> This solver is a pure Jax re-implementation of OSQP algorithm.

Reviewer: ```suggestion
This solver is a pure JAX re-implementation of the OSQP algorithm.
```

> It is jittable, supports pytrees and matvecs, but the precision is usually lower than :class:`CvxpyQP`.

Reviewer: ```suggestion
It is jittable, supports pytrees and matvecs, but the precision is usually lower than :class:`CvxpyQP` when run in float32 precision.
```

.. autosummary::
  :toctree: _autosummary

    jaxopt.OSQP

Example::

  from jaxopt import OSQP

  Q = 2 * jnp.array([[2.0, 0.5], [0.5, 1]])
  c = jnp.array([1.0, 1.0])
  A = jnp.array([[1.0, 1.0]])
  b = jnp.array([1.0])
  G = jnp.array([[-1.0, 0.0], [0.0, -1.0]])
  h = jnp.array([0.0, 0.0])

  qp = OSQP()

>   sol, state = qp.run(params_obj=(Q, c), params_eq=(A, b), params_ineq=(G, h)).params

Reviewer: Same

  print(sol.primal)
  print(sol.dual_eq)
  print(sol.dual_ineq)

See :class:`BoxOSQP` for a full description of parameters.

Box constraints QPs
-------------------

The problem takes the form:

.. math::

    \min_{x,z} \frac{1}{2} x^\top Q x + c^\top x \textrm{ subject to } A x = z, l \le z \le u

.. autosummary::
  :toctree: _autosummary

    jaxopt.BoxOSQP

> :class:`BoxOSQP` is similar to :class:`OSQP` but accepts this input format instead.

Reviewer: ```suggestion
:class:`BoxOSQP` is similar to :class:`OSQP` but accepts problems in the above box-constrained format instead.
```

The bounds ``u`` (resp. ``l``) can be set to ``inf`` (resp. ``-inf``) if required.

> 

Reviewer: Remove line

Example::

  from jaxopt import BoxOSQP

  Q = 2 * jnp.array([[2.0, 0.5], [0.5, 1]])
  c = jnp.array([1.0, 1.0])
  A = jnp.array([[1.0, 1.0], [-1.0, 0.0], [0.0, -1.0]])
  l = jnp.array([1.0, -jnp.inf, -jnp.inf])
  u = jnp.array([1.0, 0.0, 0.0])

  qp = BoxOSQP()
  sol, state = qp.run(params_obj=(Q, c), params_eq=A, params_ineq=(l, u)).params

  print(sol.primal)
  print(sol.dual_eq)
  print(sol.dual_ineq)

If required the algorithm can be speed up by setting ``check_primal_dual_infeasability`` to ``False``,
by setting ``eq_qp_preconditioner`` to ``"jacobi"`` (when possible).  
  
.. note::

  The ``tol`` parameter is used as stopping criterion when compared to primal and dual residuals.
  For over constrained problems, or badly scaled matrices, the residuals can be high, and the solver might
  never converge up to the desired accuracy.

>   In this case it is better to increases ``tol`` or to reduce ``maxiter``.  

Reviewer: Not sure about `tol`. I think the best advice is to just control `maxiter` in this case

Unconstrained QPs
-----------------

.. autosummary::
  :toctree: _autosummary

>     jaxopt.linear_solve.solve_cg

Reviewer: I think it's better to not list it, as it's a tool to solve an unconstrained QP but not a QP solver per se.

The problem takes the form:

.. math::

    \min_{x} \frac{1}{2} x^\top Q x + c^\top x

The optimality condition rewrites :math:`\nabla_x \frac{1}{2} x^\top Q x + c^\top x=Qx+c=0`.  
  

> Since the matrix :math:`Q` is assumed PSD, the best algorithm to solve :math:`Qx=-c` is *conjugate gradient*.

Reviewer: While we're at it, maybe we should write a small code snippet?

> QuadraticProgramming (deprecated)

Reviewer: I would remove all this since we don't want any more people to use it.
